File: app/src/main/java/com/hujiayucc/hook/data/AnnotationScanner.java
package com.hujiayucc.hook.data;

import android.content.Context;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Constructor;
import java.util.Enumeration;
import java.util.HashSet;
import java.util.Set;

import dalvik.system.DexFile;

public class AnnotationScanner {

    /**
     * 扫描指定包名下包含特定注解的类
     *
     * @param context         上下文
     * @param targetPackage   要扫描的包名（如："com.hujiayucc.xxx"）
     * @param annotationClass 要查找的注解类型（如：Annotation.class或Annotation::class.java）
     * @return 包含注解的类集合（Class对象）
     */
    public static Set<Class<?>> scanClassesWithAnnotation(
            Context context,
            String targetPackage,
            Class<? extends Annotation> annotationClass
    ) {
        Set<Class<?>> result = new HashSet<>();
        try {
            // 1. 获取当前应用的ClassLoader
            ClassLoader classLoader = context.getClassLoader();

            // 2. 获取APK文件路径并加载Dex文件
            String apkPath = context.getApplicationInfo().sourceDir;
            DexFile dexFile = DexFile.loadDex(apkPath, null, 0);

            // 3. 遍历Dex文件中的所有类名
            Enumeration<String> entries = dexFile.entries();
            while (entries.hasMoreElements()) {
                String className = entries.nextElement();

                // 4. 过滤目标包名下的类
                if (!className.startsWith(targetPackage)) continue;

                try {
                    // 5. 加载类对象（不初始化）
                    Class<?> clazz = Class.forName(className, false, classLoader);

                    // 6. 检查类是否包含指定注解
                    if (clazz.isAnnotationPresent(annotationClass)) {
                        result.add(clazz);
                    }

                    // 7. 检查构造函数的注解
                    for (Constructor<?> constructor : clazz.getDeclaredConstructors()) {
                        if (constructor.isAnnotationPresent(annotationClass)) {
                            result.add(clazz);
                            break; // 避免重复添加
                        }
                    }
                } catch (ClassNotFoundException | NoClassDefFoundError e) {
                    // 忽略加载失败的类
                }
            }
            dexFile.close();
        } catch (IOException e) {
            throw new RuntimeException("Dex文件扫描失败", e);
        }
        return result;
    }
}

File: app/src/main/java/com/hujiayucc/hook/utils/AnnotationScanner.kt
package com.hujiayucc.hook.utils

import android.annotation.SuppressLint
import android.content.Context
import dalvik.system.BaseDexClassLoader
import dalvik.system.DexFile
import java.lang.reflect.Field
import java.util.concurrent.ConcurrentHashMap

object AnnotationScanner {
    private data class CacheKey(
        val classLoaderIdentity: Int,
        val targetPackage: String,
        val annotationNames: String
    )

    private data class FieldKey(
        val clazz: Class<*>,
        val name: String
    )

    private val classNameCache = ConcurrentHashMap<CacheKey, Set<String>>()
    private val fieldCache = ConcurrentHashMap<FieldKey, Field>()

    /**
     * 扫描指定包名下包含特定注解的类
     *
     * @param context         上下文
     * @param targetPackage   要扫描的包名（如："xxx.xxx.xxx"）
     * @param annotationClasses 要查找的注解类型（如：Annotation.class或Annotation::class.java）
     * @return 包含注解的类集合（Class对象）
     */
    fun scanClassesWithAnnotation(
        context: Context,
        targetPackage: String,
        annotationClasses: Collection<Class<out Annotation>>
    ): Set<Class<*>> {
        return scanClassesWithAnnotation(context.classLoader, targetPackage, annotationClasses)
    }

    /**
     * 扫描指定包名下包含特定注解的类
     *
     * @param classLoader     类加载器
     * @param targetPackage   要扫描的包名（如："xxx.xxx.xxx"）
     * @param annotationClasses 要查找的注解类型（如：Annotation.class或Annotation::class.java）
     * @return 包含注解的类集合（Class对象）
     */
    fun scanClassesWithAnnotation(
        classLoader: ClassLoader,
        targetPackage: String,
        annotationClasses: Collection<Class<out Annotation>>
    ): Set<Class<*>> {
        if (annotationClasses.isEmpty()) return emptySet()

        val annotationNameSet = annotationClasses.asSequence()
            .map { it.name }
            .toSet()
        val annotationKey = annotationNameSet.asSequence()
            .sorted()
            .joinToString(separator = "|")
        val cacheKey = CacheKey(System.identityHashCode(classLoader), targetPackage, annotationKey)

        classNameCache[cacheKey]?.let { cachedClassNames ->
            return loadClassesByName(classLoader, cachedClassNames)
        }

        val matchedClasses = scanMatchedClasses(classLoader, targetPackage, annotationNameSet)
        classNameCache[cacheKey] = matchedClasses.asSequence().map { it.name }.toCollection(LinkedHashSet())
        return matchedClasses
    }

    @Suppress("DEPRECATION")
    private fun scanMatchedClasses(
        classLoader: ClassLoader,
        targetPackage: String,
        annotationNames: Set<String>
    ): Set<Class<*>> {
        val result = LinkedHashSet<Class<*>>()
        extractDexFiles(classLoader).forEach { dexFile ->
            val entries = dexFile.entries()
            while (entries.hasMoreElements()) {
                val className = entries.nextElement()
                if (!className.startsWith(targetPackage)) continue

                val clazz = loadClassOrNull(classLoader, className) ?: continue
                if (clazz.hasAnyTargetAnnotation(annotationNames)) result.add(clazz)
            }
        }
        return result
    }

    private fun loadClassesByName(classLoader: ClassLoader, classNames: Set<String>): Set<Class<*>> {
        val result = LinkedHashSet<Class<*>>(classNames.size)
        classNames.forEach { className ->
            loadClassOrNull(classLoader, className)?.let { result.add(it) }
        }
        return result
    }

    private fun loadClassOrNull(classLoader: ClassLoader, className: String): Class<*>? {
        return try {
            Class.forName(className, false, classLoader)
        } catch (_: Throwable) {
            null
        }
    }

    private fun Class<*>.hasAnyTargetAnnotation(annotationNames: Set<String>): Boolean {
        if (annotations.any { it.annotationClass.java.name in annotationNames }) return true
        return declaredConstructors.any { constructor ->
            constructor.annotations.any { it.annotationClass.java.name in annotationNames }
        }
    }

    private fun extractDexFiles(classLoader: ClassLoader): List<DexFile> {
        if (classLoader !is BaseDexClassLoader) {
            return emptyList()
        }
        return try {
            @SuppressLint("DiscouragedPrivateApi")
            val dexPathList = cachedDeclaredField(BaseDexClassLoader::class.java, "pathList").get(classLoader)
            val dexElements = cachedDeclaredField(dexPathList.javaClass, "dexElements").get(dexPathList) as Array<*>
            dexElements.mapNotNull { element ->
                element ?: return@mapNotNull null
                cachedDeclaredField(element.javaClass, "dexFile").get(element) as? DexFile
            }
        } catch (_: Throwable) {
            emptyList()
        }
    }

    private fun cachedDeclaredField(clazz: Class<*>, name: String): Field {
        val key = FieldKey(clazz, name)
        return fieldCache.getOrPut(key) {
            clazz.getDeclaredField(name).apply { isAccessible = true }
        }
    }
}